### PR TITLE
Fold output of each benchmark

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

This PR improves output of benchmarks run on GitHub Actions CI to group and fold output for each benchmark.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md)
- [x] Benchmarks output improved
    - [x] [engine run](https://github.com/enso-org/enso/actions/runs/10627416248)
    - [x] [stdlib run](https://github.com/enso-org/enso/actions/runs/10627603184)

### CODE REVIEW

File: lib/java/benchmarks-common/src/main/java/org/enso/interpreter/bench/BenchmarksRunner.java [hunk 1]
```diff
@@ -107,4 +117,76 @@ private static BenchmarkItem reportResult(String label, RunResult result) throws
...
+
+    @Override
+    public void iterationResult(
+        BenchmarkParams benchParams, IterationParams params, int iteration, IterationResult data) {
+      output.iterationResult(benchParams, params, iteration, data);
+    }
+
+    @Override
+    public void startBenchmark(BenchmarkParams benchParams) {
+      output.println("::group::" + benchParams.getBenchmark());
```
Comment (JaroslavTulach) [inline lib/java/benchmarks-common/src/main/java/org/enso/interpreter/bench/BenchmarksRunner.java:150]: The `::group::` format seems to be mentioned at [this discussion](https://github.com/orgs/community/discussions/25618).

**Comment by JaroslavTulach** [comment]
**Update**: solved in 9ba4428

I don't know how to print plain line without any prefix in our Rust build system. If I use `info!` like in 37d0385 the line gets prefixed and looses its `::` command prefix. If I use `println!` like in b1ef6fb the text isn't displayed in the log at all! Any advice from Rust experts?

**Review by Akirathan — DISMISSED** [review]
This is a good idea. I am interested in how the end result will look like. I had no idea that there is some output grouping for GH actions. We could probably integrate this functionality even for other actions.

**Comment by JaroslavTulach** [comment]
Locally the _command lines_ seem to be determined fine. If I execute, I see:
```bash
enso$ ./run backend benchmark runtime --minimal-run
 INFO ide_ci::program::command: sbt ℹ️ [info] # *** WARNING: Use non-forked runs only for debugging purposes, not for actual performance runs. ***
 INFO ide_ci::program::command: sbt ℹ️ [info] Iteration   1: 102.137 ms/op
::endgroup::
::group::org.enso.interpreter.bench.benchmarks.semantic.ArrayProxyBenchmarks.sumOverComputingProxy
 INFO ide_ci::program::command: sbt ℹ️ [info] # Run progress: 12.68% complete, ETA 00:01:29
 INFO ide_ci::program::command: sbt ℹ️ [info] # Fork: N/A, test runs in the host VM
```
e.g. both `::endgroup` and `::group` are properly detected and the "INFO..." prefix isn't printed. Let's see [what does the CI print](https://github.com/enso-org/enso/actions/runs/10627416248)?

folds

Finally it does [something reasonable](https://github.com/enso-org/enso/actions/runs/10627416248/job/29460495368#step:7:1423)!

**Comment by JaroslavTulach** [comment]
Certain parts of the [stdlibs bench run output](https://github.com/enso-org/enso/actions/runs/10627603184/job/29460978326#step:7:6082) look nice as well.

Some messages "escape" the grouping. Not sure why and whether that matters that much to prevent integration. _Hypothesis_:
- either [output prints to stdout](https://github.com/enso-org/enso/pull/10914/files#diff-89d87e0a1a41641d93315285f61ed0538fe5fbddd506255c8bd7fe7618c3977aR142) and [compilation detection](https://github.com/enso-org/enso/pull/10574/files#diff-8c21fff89189c63761ac819eff6f55a53c500835ca7a5da5d427ea1e8c2da735R350) prints to `System.err` and the output gets mixed
- or two two messages are printed by different threads

Let's try to [switch to stdout](https://github.com/enso-org/enso/pull/10914/commits/c698ef70a681823184880217cc331fdd12403762) - [stdlib benchmark](https://github.com/enso-org/enso/actions/runs/10628829078) is running.

**Update**: The [latest stdlibs output](https://github.com/enso-org/enso/actions/runs/10628829078/job/29464422212#step:7:1381) looks pretty decent. Looks like c698ef7 helped.

File: build/ci_utils/src/program/command.rs [hunk 2]
```diff
@@ -469,7 +469,22 @@ pub fn spawn_log_processor(
...
-                        info!("{prefix} {line}");
+                        let mut command = false;
+                        if let Some(command_at) = line.find("::") {
+                            if let Some(group_at) = line.find("group") {
+                                // we support:
+                                // ::group
+                                // ::endgroup
+                                if command_at < group_at && group_at < command_at + 10 {
+                                    let line_without_prefix = &line[command_at..];
+                                    println!("{line_without_prefix}");
```
Comment (4e6) [inline build/ci_utils/src/program/command.rs:479]: cleanup

**Review by 4e6 — APPROVED** [review]

**Review by Akirathan — CHANGES_REQUESTED** [review]
After some thinking, I believe that your changes are unnecessary, and seeing the [current folded output](https://github.com/enso-org/enso/actions/runs/10627416248/job/29460495368), it is very messy - The group for [benchGenerateList](https://github.com/enso-org/enso/actions/runs/10627416248/job/29460495368#step:7:1475) obviously contains output even from some other benchmark for example.

May I ask you again to explain your use-case that this folding will solve, and why do you think it is needed?

Note that at the end of each benchmark run (the very end of the whole job output) there is table-like structure with all the scores. In your case it is at https://github.com/enso-org/enso/actions/runs/10627416248/job/29460495368#step:7:1886. If you want to see the output for a single benchmark, you can just type its name in the search field. You would need to do that even if the output was folded.

**EDIT**: If you still think the folding will be beneficial, I will revoke my PR reject after I see that the folding works and that no output is assigned to a wrong group.

File: build/ci_utils/src/program/command.rs (hunk 2 above)
Comment (4e6) [reply build/ci_utils/src/program/command.rs:479]: Ok, maybe there is a comment needed that `println` is used to print the `::group` mark without the prefix

**Comment by JaroslavTulach** [comment]
> ... changes are unnecessary...

I am sad you have changed your mind. I believe that with few more _homey touches_, we can make this folding work well and increase overall satisfaction.

> 
> May I ask you again to explain your use-case that this folding will solve, and why do you think it is needed?

- I want to hide non-interesting crap - folding would achieve that
- I want to organize things per interest - using title of a benchmark for navigation seems human friendly
- I was hoping we tackle the size of the log (didn't work trying more in 2344fac) as we are reaching a limit on stdlibs benchmarks

> Note that at the end of each benchmark run (the very end of the whole job output) there is table-like structure with all the scores. In your case it is at https://github.com/enso-org/enso/actions/runs/10627416248/job/29460495368#step:7:1886. 

Right now the log stops at [22% of stdlibs benchmark run](https://github.com/enso-org/enso/actions/runs/10628829078/job/29464422212#step:7:15982) so the table isn't really visible.

> If you want to see the output for a single benchmark, you can just type its name in the search field. You would need to do that even if the output was folded.

I've just tried and folding has _no negative effect on searching_. These `::group:` folds are automatically expanded when the found text is in a collapsed fold.

> If you still think the folding will be beneficial

The [latest stdlibs output](https://github.com/enso-org/enso/actions/runs/10628829078/job/29464422212#step:7:1381) looks pretty decent. Looks like c698ef7 helped.

File: build/ci_utils/src/program/command.rs (hunk 2 above)
Comment (JaroslavTulach) [reply build/ci_utils/src/program/command.rs:479]: done in 4c0ddd51efb60df3b74c76c5ca8de70e92c18509

File: lib/java/benchmarks-common/src/main/java/org/enso/interpreter/bench/BenchmarksRunner.java (hunk 1 above)
Comment (radeusgd) [reply lib/java/benchmarks-common/src/main/java/org/enso/interpreter/bench/BenchmarksRunner.java:150]: It is documented here: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#grouping-log-lines

Although the documentation is not very detailed.

**Comment by JaroslavTulach** [comment]
The newest idea c36fdad to solve the _too huge log issue_ is to differentiate between execution of all benchmarks and execution of a selected benchmark(s):
- running `bench` limits the output to ten lines per benchmark - let's see [if that helps](https://github.com/enso-org/enso/actions/runs/10737648553)
- running `benchOnly Xyz` dumps full output

As a result the CI run will be less verbose, but one will have a chance to re-run individual benchmark locally and see whole output.

**Comment by JaroslavTulach** [comment]
> at the end of each benchmark run (the very end of the whole job output) there is table-like structure with all the scores. 

Yes, with the [latest run the table is finally visible](https://github.com/enso-org/enso/actions/runs/10737648553/job/29779559241#step:7:9461).

>  you can just type its name in the search field. 

That is still possible even the output is folded.

> If you still think the folding will be beneficial, I will revoke my PR reject 

Yes, consider revoking your reject.

**Review by Akirathan — APPROVED** [review]
The [latest output](https://github.com/enso-org/enso/actions/runs/10737648553/job/29779559241#step:7:9461) seems reasonable. The fact that when all the benchmarks are run, the output is less verbose is good - it helps to be able to interactively browse the output on the CI.

**Review by jdunkerley — APPROVED** [review]

**Pull request merged**